src/tools/rbd_mirror/ImageDeleter.cc

```diff
@@ -0,0 +1,367 @@
+  dout(20) << "enter" << dendl;
+  while(m_running.read()) {
+    m_delete_queue_lock.Lock();
+    while (m_delete_queue.empty()) {
+      dout(20) << "waiting for delete requests" << dendl;
+      m_delete_queue_cond.Wait(m_delete_queue_lock);
+
+      if (!m_running.read()) {
+        m_delete_queue_lock.Unlock();
+        dout(20) << "return" << dendl;
+        return;
+      }
+    }
+
+    unique_ptr<DeleteInfo> del_info = std::move(m_delete_queue.front());
+    m_delete_queue.pop();
+    m_delete_queue_lock.Unlock();
+
+    process_image_delete(std::move(del_info));
+  }
+}
+
+void ImageDeleter::schedule_image_delete(uint64_t local_pool_id,
+                                         const std::string& local_image_id,
+                                         const std::string& global_image_id) {
+  dout(20) << "enter" << dendl;
+  Mutex::Locker l(m_delete_queue_lock);
+  m_delete_queue.push(unique_ptr<DeleteInfo>(
+        new DeleteInfo(local_pool_id, local_image_id, global_image_id)));
+  m_delete_queue_cond.Signal();
+}
+
+void ImageDeleter::process_image_delete(unique_ptr<DeleteInfo> del_info,
+                                        FailedStep step, int retries) {
+
+  std::string del_info_str = del_info->to_string();
+  dout(20) << "start processing delete request: " << del_info_str << dendl;
+  int r;
+  cls::rbd::MirrorImage mirror_image;
+
+  // remote image was disabled, now we need to delete local image
+  IoCtx ioctx;
+  r = m_local->ioctx_create2(del_info->local_pool_id, ioctx);
+  if (r < 0) {
+    derr << "error accessing local pool: " << cpp_strerror(r) << dendl;
+    enqueue_failed_delete(std::move(del_info), FAILED_ON_DISABLE_MIRRORING, r,
+                          retries);
+    return;
+  }
+
+  bool has_snapshots;
+  bool has_children;
+  r = image_has_snapshots_and_children(&ioctx, del_info->local_image_id,
+                                       &has_snapshots, &has_children);
+  if (r < 0) {
+    enqueue_failed_delete(std::move(del_info), FAILED_ON_DISABLE_MIRRORING, r,
+                          retries);
+    return;
+  }
+
+  if (has_children) {
+    // there are still clones of snapshots of this image, therefore send the
+    // delete request to the end of the queue
+    dout(20) << "local image id " << del_info->local_image_id << " has "
+             << "snapshots with cloned children, postponing deletion..."
+             << dendl;
+    Mutex::Locker l(m_delete_queue_lock);
+    m_delete_queue.push(std::move(del_info));
+    return;
+  }
+
+  string img_name;
+  r = cls_client::dir_get_name(&ioctx, RBD_DIRECTORY, del_info->local_image_id,
+                               &img_name);
+  if (r < 0) {
+    derr << "error retrieving local image name for image id "
+         << del_info->local_image_id << ": " << cpp_strerror(r) << dendl;
+    enqueue_failed_delete(std::move(del_info), FAILED_ON_DISABLE_MIRRORING, r,
+                          retries);
+    return;
+  }
+
+  switch (step) {
```

Comment:
Why worry about tracking the state?  You should be able to just fall right through w/o worry (i.e. if mirroring has already been disabled, it's a no-op).